### PR TITLE
Add Github actions support

### PR DESCRIPTION
**Pull request by AlanFoster**
Adds Github actions support to pcaprub; Issue spotted as part of https://github.com/pcaprub/pcaprub/pull/61

After being merged, the pcaprub gem will be built on Windows/Ubuntu/macOS:

Screenshot 2023-03-22 233956

Example of tests passing:

Screenshot 2023-03-22 234110

### CODE REVIEW

File: test/test_pcaprub_unit.rb [hunk 1]
```diff
@@ -122,7 +122,7 @@ def test_create_from_primitives
 
   def test_set_datalink
     d = Pcap.lookupdev
-    o = Pcap.open_live(d, 65535, true, -1)
+    o = Pcap.open_live(d, 65535, true, 1)
```
Comment (AlanFoster) [inline test/test_pcaprub_unit.rb:125]: Fixes the following issue on macOS:

```
E
===============================================================================
Error: test_set_datalink(PCAPRUB::Pcap::UnitTest): RuntimeError: en0: BIOCSRTIMEOUT: Unknow
rake aborted!
/Users/runner/work/pcaprub/pcaprub/test/test_pcaprub_unit.rb:125:in `open_live'
Command failed with status (1): [ruby -w -I"lib:lib:test" /Users/runner/hostedtoolcache/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb "test/test_pcaprub_unit.rb" ]
/Users/runner/work/pcaprub/pcaprub/test/test_pcaprub_unit.rb:125:in `test_set_datalink'
/Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle:23:in `load'
     122:
/Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle:23:in `<main>'
     123:   def test_set_datalink
Tasks: TOP => default => test
     124:     d = Pcap.lookupdev
(See full trace by running task with --trace)
  => 125:     o = Pcap.open_live(d, 65535, true, -1)
     126:     dls = o.listdatalinks
     127:     begin
     128:       assert_equal(o,o.setdatalink(dls.values.first))
===============================================================================
.
```

From a quick Google it looked like `-1` as a timeout was invalid; And all of the other `Pcap.open_live` calls are set to `1`, so I've updated this here too for consistency. The macOS tests pass after this change.

File: .github/workflows/verify.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          if (!(Test-Path -Path C:\WpdPack\Lib\x64\wpcap.lib)) {
+            throw "failed extracting wpcap"
+          }
+
+      - name: build
+        run: |
+          bundle install
+          rake gem
+
+      # Skipped for now as the tests do not pass on Windows
```
Comment (AlanFoster) [inline .github/workflows/verify.yml:77]: The windows tests do not work, either on CI or a fresh Windows box:

```
C:/hostedtoolcache/windows/Ruby/3.0.2/x64/bin/ruby.exe -w -I"lib;lib;test" C:/hostedtoolcache/windows/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb "test/test_pcaprub_unit.rb" 
D:/a/pcaprub/pcaprub/lib/pcaprub/ext.rb:14:in `require': 126: The specified module could not be found.   - D:/a/pcaprub/pcaprub/lib/pcaprub_c.so (LoadError)
	from D:/a/pcaprub/pcaprub/lib/pcaprub/ext.rb:14:in `rescue in <top (required)>'
	from D:/a/pcaprub/pcaprub/lib/pcaprub/ext.rb:1:in `<top (required)>'
	from D:/a/pcaprub/pcaprub/lib/pcaprub.rb:5:in `require'
	from D:/a/pcaprub/pcaprub/lib/pcaprub.rb:5:in `<module:PCAPRUB>'
	from D:/a/pcaprub/pcaprub/lib/pcaprub.rb:1:in `<top (required)>'
	from D:/a/pcaprub/pcaprub/test/test_helper.rb:9:in `require'
	from D:/a/pcaprub/pcaprub/test/test_helper.rb:9:in `<top (required)>'
	from D:/a/pcaprub/pcaprub/test/test_pcaprub_unit.rb:7:in `require'
	from D:/a/pcaprub/pcaprub/test/test_pcaprub_unit.rb:7:in `<top (required)>'
	from C:/hostedtoolcache/windows/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `require'
	from C:/hostedtoolcache/windows/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
	from C:/hostedtoolcache/windows/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
	from C:/hostedtoolcache/windows/Ruby/3.0.2/x64/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
```

**Comment by AlanFoster** [comment]
cc @crondaemon - hopefully this is what you were after; I can potentially look more into the failing Windows tests, but I don't know if they've ever passed

File: test/test_pcaprub_unit.rb [hunk 3]
```diff
@@ -133,7 +133,7 @@ def test_set_datalink
   end
 
   def test_monitor
-    return if RUBY_PLATFORM =~ /mingw|win/
+    return if RUBY_PLATFORM =~ /mingw/
```
Comment (AlanFoster) [inline test/test_pcaprub_unit.rb:136]: This had a false positive on skipping this test on macOS, as the RUBY_PLATFORM was `x86_64-darwin19`

File: test/test_pcaprub_unit.rb [hunk 4]
```diff
@@ -164,7 +164,11 @@ def test_filter
 
   def test_lib_version
     v = Pcap.lib_version.split
-    assert_equal "libpcap", v[0]
+    if RUBY_PLATFORM =~ /mingw/
+      assert_equal "WinPcap", v[0]
```
Comment (AlanFoster) [inline test/test_pcaprub_unit.rb:168]: This test failed on Windows due to the name difference:

image

File: script/installwinpcap.ahk
Comment (AlanFoster) [inline script/installwinpcap.ahk:1]: It turns out that `choco install winpcap` doesn't work due to outdated AHK1 file not being compatible with AutoHotkey v2; I'll try and contribute a separate fix upstream to choco so this file can be deleted in the future, but it shouldn't be a blocker for now.

**Comment by AlanFoster** [comment]
@crondaemon Updated to work now on Windows now, all three platforms should be building/testing as expected :+1:

<details>
<summary>Option details on how I debugged the windows failures</summary>

When I tried to run `bundle exec rake` I got the following error:
```
C:/pcaprub/lib/pcaprub/ext.rb:16:in `require': 126: The specified module could not be found.   - C:/pcaprub/lib/pcaprub_c.so (LoadError)
        from C:/pcaprub/lib/pcaprub/ext.rb:16:in `rescue in <top (required)>'
        from C:/pcaprub/lib/pcaprub/ext.rb:1:in `<top (required)>'
        from C:/pcaprub/lib/pcaprub.rb:5:in `require'
        from C:/pcaprub/lib/pcaprub.rb:5:in `<module:PCAPRUB>'
        from C:/pcaprub/lib/pcaprub.rb:1:in `<top (required)>'
        from C:/pcaprub/test/test_helper.rb:9:in `require'
        from C:/pcaprub/test/test_helper.rb:9:in `<top (required)>'
        from C:/pcaprub/test/test_pcaprub_unit.rb:7:in `require'
        from C:/pcaprub/test/test_pcaprub_unit.rb:7:in `<top (required)>'
        from C:/tools/ruby30/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `require'
        from C:/tools/ruby30/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
        from C:/tools/ruby30/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
        from C:/tools/ruby30/lib/ruby/gems/3.0.0/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
```

I confirmed the `.so` file was being compiled correctly and existed on disk; so I checked out the dependencies:

```
c:\pcaprub>dumpbin /dependents c:\pcaprub\lib\pcaprub_c.so
Microsoft (R) COFF/PE Dumper Version 14.29.30148.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file c:\pcaprub\lib\pcaprub_c.dll

File Type: DLL

  Image has the following dependencies:

    KERNEL32.dll
    msvcrt.dll
    WS2_32.dll
    x64-msvcrt-ruby300.dll
    wpcap.dll
```

Then I ran procmon to see which dll wasn't being loaded, first adding the filters for missing dlls:

image

Then I ran the test suite again to catch the missing library:

image

It looked like `wpcap.dll` is required at runtime for the tests to work. and the `wpdpack` files were only required for compilation. I tried using `choco install winpcap` to install this, but it turns out the AHK file is outdated and doesn't work. I've got a solution created for this PR, and I'll try to contribute it upstream.

</details>

**Comment by crondaemon** [comment]
Github actions are failing. Did you remove the branch?

**Comment by AlanFoster** [comment]
@crondaemon I feel like that was a github UI glitch; The action's result should not have appeared in this repo until they're merged into this repo. I believe Github actions clean up automatically after $x weeks of being run, which might've caused that glitch.

I've pushed the code again to trigger a build, and the latest build result is here: https://github.com/AlanFoster/pcaprub/actions/runs/4759578018

**Comment by crondaemon** [comment]
This PR is no longer necessary, then.

**Comment by AlanFoster** [comment]
@crondaemon I think there's been a mis-communication; this PR adds support for Github actions and has not been deleted. 

> crondaemon: Github actions are failing. 

They are passing here: https://github.com/AlanFoster/pcaprub/actions/runs/4759578018 - the error you saw just a bug in Github's UI

**Comment by crondaemon** [comment]
I've reopened the PR. If you fix the CI I will merge it. Later on I will add support for more ruby versions.

**Comment by crondaemon** [comment]
> I've reopened the PR. If you fix the CI I will merge it. Later on I will add support for more ruby versions.

If I got your point, the CI fails because of permissions, and will succeed once merged, correct?

**Pull request merged**